Add suctl diagnose in cluster test

tomer-shefler opened the pull request:
Depends on https://github.com/submariner-io/submariner-operator/pull/2283
Depends on https://github.com/submariner-io/submariner-operator/pull/2288

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

tomer-shefler committed:
Add subctl diagnose in-cluster test

Signed-off-by: Tomer Shefler <sheflertomer@gmail.com>
Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,35 @@ test_subctl_gather
 _subctl diagnose firewall inter-cluster --validation-timeout 20 "${KUBECONFIGS_DIR}"/kind-config-cluster1 "${KUBECONFIGS_DIR}"/kind-config-cluster2
 
+# Test subctl diagnose in-cluster
+cat <<EOF | kubectl apply -f -
```
```suggestion
kubectl apply -f - <<EOF
```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,35 @@ test_subctl_gather
+EOF
+kubectl wait --for=condition=complete job/submariner-diagnose -n submariner-operator --timeout 120s
+retVal=$?
+if [ $retVal -ne 0 ]; then
```
```suggestion

kubectl wait --for=condition=complete job/submariner-diagnose -n "$subm_ns"  --timeout 120s &
kubectl wait --for=condition=failed job/submariner-diagnose -n "$subm_ns"  --timeout 120s && exit 1 &
if ! wait -n; then
    kill %1 %2 2> /dev/null
```

This avoids waiting for 2 minutes for it to time out in case of failure.

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,35 @@ test_subctl_gather
+    echo "Test failed: printing logs"
+    kubectl logs job/submariner-diagnose -n submariner-operator
+fi
+kubectl delete job submariner-diagnose -n submariner-operator
```
```suggestion

kubectl delete job submariner-diagnose -n "$subm_ns"
echo "::endgroup::"

```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,35 @@ test_subctl_gather
+      containers:
+      - name: submariner-diagnose
+        image: quay.io/submariner/subctl:devel
+        command: ["subctl",  "diagnose", "all", "--in-cluster"]
```
```suggestion
        command: ["subctl",  "diagnose", "all", "--in-cluster", "--validation-timeout", "20"]
```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,35 @@ test_subctl_gather
+kind: Job
+metadata:
+  name: submariner-diagnose
+  namespace: submariner-operator
```
```suggestion
  namespace: ${subm_ns}
```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,35 @@ test_subctl_gather
+    spec:
+      containers:
+      - name: submariner-diagnose
+        image: quay.io/submariner/subctl:devel
```
```suggestion
        image: localhost:5000/subctl:local
```

You'll also need to add `subctl` to `PRELOAD_IMAGES` - https://github.com/submariner-io/subctl/blob/devel/Makefile#L27

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,35 @@ test_subctl_gather
+retVal=$?
+if [ $retVal -ne 0 ]; then
+    echo "Test failed: printing logs"
+    kubectl logs job/submariner-diagnose -n submariner-operator
```
```suggestion
    kubectl logs job/submariner-diagnose -n "$subm_ns"
```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,35 @@ test_subctl_gather
 _subctl diagnose all --validation-timeout 20
 _subctl diagnose firewall inter-cluster --validation-timeout 20 "${KUBECONFIGS_DIR}"/kind-config-cluster1 "${KUBECONFIGS_DIR}"/kind-config-cluster2
 
+# Test subctl diagnose in-cluster
```
```suggestion
# Test subctl diagnose in-cluster
echo "::group::Validating 'subctl diagnose' in-cluster"
```

This groups everything in CI to make it easier to read

mkolesnik reviewed (changes requested):
Thanks, a few comments

mkolesnik commented:
Opened #245 for the failure

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,41 @@ test_subctl_gather
+kubectl wait --for=condition=complete job/submariner-diagnose -n "${subm_ns}"  --timeout 120s & 
+kubectl wait --for=condition=failed job/submariner-diagnose -n "${subm_ns}"  --timeout 120s && exit 1 &
+if ! wait -n; then
+	  kill %1 %2 2> /dev/null
```
```suggestion
    kill %1 %2 2> /dev/null
```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,41 @@ test_subctl_gather
+	  kill %1 %2 2> /dev/null
+    echo "Test failed: printing logs"
+    kubectl logs job/submariner-diagnose -n ${subm_ns}
+fi
```
```suggestion
fi

```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,41 @@ test_subctl_gather
+if ! wait -n; then
+	  kill %1 %2 2> /dev/null
+    echo "Test failed: printing logs"
+    kubectl logs job/submariner-diagnose -n ${subm_ns}
```
As we discussed, it's better to always print the logs.

Try running it right after apply with `-f` flag, and after the log exists you can check the job, which should be done by then, so we might not need all the `wait` logic

mkolesnik replied:
Also don't forget to exit when the job fails, so that CI shows the failure

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,43 @@ test_subctl_gather
+echo "Print diagnose logs"
+kubectl logs job/${job_name} -n ${subm_ns} -f
+if kubectl get jobs ${job_name} -n ${subm_ns} -o jsonpath='{.status.conditions[?(@.type=="Failed")].status}' | grep "True" > /dev/null
+then
```
```suggestion
if kubectl get jobs "${job_name}" -n "${subm_ns}" -o jsonpath='{.status.conditions[?(@.type=="Failed")].status}' | grep -q "True"; then
```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,43 @@ test_subctl_gather
+kubectl logs job/${job_name} -n ${subm_ns} -f
+if kubectl get jobs ${job_name} -n ${subm_ns} -o jsonpath='{.status.conditions[?(@.type=="Failed")].status}' | grep "True" > /dev/null
+then
+  kubectl delete job ${job_name} -n ${subm_ns}
+  exit 1
```
```suggestion
    kubectl delete job "${job_name}" -n "${subm_ns}"
    exit 1
```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,43 @@ test_subctl_gather
+  exit 1
+fi
+
+kubectl delete job ${job_name} -n ${subm_ns}
```
```suggestion
kubectl delete job "${job_name}" -n "${subm_ns}"
```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,43 @@ test_subctl_gather
+EOF
+
+echo "Print diagnose logs"
+kubectl logs job/${job_name} -n ${subm_ns} -f
```
```suggestion
kubectl logs "job/${job_name}" -n "${subm_ns}" -f
```

mkolesnik reviewed (commented):
Thanks, a few nit-picks but overall looks good.

mkolesnik commented:
Hmm seems that youll need to wait for container to be created anyhow before querying the logs - https://github.com/submariner-io/subctl/runs/8307689456?check_suite_focus=true#step:3:6095

nyechiel commented:
@tomer-shefler can you resolve the conflict please? also note that the PR title has a typo (suctl instead of `subctl`)

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,46 @@ test_subctl_gather
+EOF
+
+echo "Wait for job to be ready"
+kubectl wait --for=condition=Ready "job/${job_name}"
```
```suggestion
kubectl wait --for=condition=Ready -n "$subm_ns" "job/${job_name}"
```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,46 @@ test_subctl_gather
+
+echo "Print diagnose logs"
+kubectl logs "job/${job_name}" -n "${subm_ns}" -f
+if kubectl get jobs "${job_name}" -n "${subm_ns}" -o jsonpath='{.status.conditions[?(@.type=="Failed")].status}' | grep "True" > /dev/null
+then
```
```suggestion
if [[ $(kubectl get jobs "${job_name}" -n "${subm_ns}" -o jsonpath='{.status.conditions[?(@.type=="Failed")].status}') = "True" ]]; then
```

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -68,6 +68,46 @@ test_subctl_gather
+kubectl logs "job/${job_name}" -n "${subm_ns}" -f
+if kubectl get jobs "${job_name}" -n "${subm_ns}" -o jsonpath='{.status.conditions[?(@.type=="Failed")].status}' | grep "True" > /dev/null
+then
+  kubectl delete job "${job_name}" -n "${subm_ns}"
+  exit 1
```
```suggestion
    kubectl delete job "${job_name}" -n "${subm_ns}"
    exit 1
```

For consistency with indentation in the file

tpantelis commented:
globalnet and service discovery need additional privs for the `submariner-diagnose` role: https://github.com/submariner-io/submariner-operator/pull/2264

mkolesnik commented on Makefile:
```diff
@@ -66,7 +66,7 @@ export PATH := $(CURDIR)/cmd/bin:$(PATH)
 # Build subctl before deploying to ensure we use that
 # (with the PATH set above)
-deploy: cmd/bin/subctl
+deploy: cmd/bin/subctl images
```
This shouldn't be necessary, why did you add it?

```suggestion
deploy: cmd/bin/subctl
```

tpantelis replied:
I ran into an issue where I made a change to `subctl` but it wasn't picked up by the `submariner-diagnose` image used for the system tests. This seemed to fix that.

mkolesnik replied:
The image is always built before deployment, no need to add this.
The problem you had was probably something else (perhaps you had stale shipyard image/files around).

mkolesnik reviewed (commented):
Looks good, one question/suggestion

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -16,7 +16,7 @@ function deploy_env_once() {
 
     # Print GHA groups to make looking at CI output easier
     printf "::group::Deploying the environment"
-    make deploy SETTINGS="$SETTINGS" using="${USING}" -o package/.image.subctl
```
This makes sense :+1:

tpantelis committed:
Use empty namespace for in-cluster

...instead of "default" to workaround the permissions issue
when running diagnose inside a pod where the cluster role doesn't
have "get" access to namespaces. The internal network pod
scheduling code defaults to "submariner-operator" if no
namespace is provided. This will be properly addressed later.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>
